specs/agents/tracing-instrumentation-aws.md

```diff
@@ -0,0 +1,53 @@
+## AWS services spans
+
+We describe how to instrument some of AWS' services in this document.
+Many of the services have a fairly small API beyond the basic administrative commands. We therefore
+only instrument the most interesting and often-repeated operations.
+
+### S3 (Simple Storage Service)
+
+AWS Simple Storage Service offers object storage via a REST API. The objects are organized into buckets, which are 
+themselves organized into regions.
+
+- `span.name`: The span name should follow this pattern: `S3 <OPERATION_NAME> <BUCKET-NAME>`. For example,
+`S3 GET_OBJECT myBucket`. Note that the operation name is in all caps with underscores (snake case in caps).
+- `span.type`: `storage`
+- `span.subtype`: `s3`
+- `span.action`: The operation name in snake case in lower case. For example ‘get_object’. 
+
+#### Span context fields
+
+- **`context.destination.address`**: optional. Not available in some cases. Only set if the actual connection is available.
+- **`context.destination.port`**: optional. Not available in some cases. Only set if the actual connection is available.
+- **`context.destination.region`**: mandatory. The AWS region where the bucket is.
+- **`context.destination.service.name`**: mandatory. Use `s3`
+- **`context.destination.service.resource`**: mandatory. The bucket ARN, if available, otherwise simply `s3/<BUCKET-NAME`.
+Note that the ARN will contain the bucket name. The s3 API allows either the bucket name or the ARN to be provided when
+referring to a bucket so use whichever the user has provided.
```

Comment:
The s3 API allows a bucket to be specified either by its ARN or the bucket name so this field may have different values for the same resource, depending on how the user uses the API. That may be an issue so perhaps we should only allow the ARN in this field.

Reply:
I wonder if we should actually parse out the bucket name and always use that? Will we have to jump through hoops to get the ARN if the user only provides the bucket name? With the ARN we can always reduce to the bucket name.

Reply:
The bucket name is only unique to a region. So I'm wondering if it would be enough to put just the bucket name in this field, as it wouldn't provide the same level of cardinality as the ARN?

Reply:
Ah, I'm wrong! It's not unique to a region. So the bucket name would be enough to identify a unique bucket.
https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingBucket.html

Reply:
Ah, nice! Then I think we should try to detect if it's an ARN and pull out just the bucket name.